### PR TITLE
[Importer] Support importing pages with authentication

### LINKED ISSUE #2365: [Importer] Support importing pages with authentication
In xwalk projects (Danaher, Comwrap, IBM …) we need to run the importer on AEM author pages which require authentication. This is currently not possible because the HelixImportServer removes the header information in [1].

One solution we discussed with @kptdobe, @Buuhuu, @bpauli, @jckautzmann is to run the import CLI with a flag (e.g. allowHeaders) pointing to a json file (e.g. headers.json)  in which the headers are defined and make the import proxy server pass the headers to the target server. For current projects, either the Cookie or the Authorisation header will be used.

[1] https://github.com/adobe/helix-cli/blob/main/src/server/HelixImportServer.js#L99

### PR DESCRIPTION
**Pull request by jckautzmann**
- add support for setting headers on the CLI

Fixes #2365 

Please ensure your pull request adheres to the following guidelines:
- [ ] make sure to link the related issues in this description
- [ ] when merging / squashing, make sure the fixed issue references are visible in the commits, for easy compilation of release notes

## Related Issues

Thanks for contributing!

### CODE REVIEW

File: src/server/HelixImportServer.js [hunk 1]
```diff
@@ -63,6 +64,20 @@ export class HelixImportServer extends BaseServer {
...
+    // eslint-disable-next-line no-underscore-dangle
+    const filePath = this._project._headersFile;
+    if (!filePath) return;
+    let cliHeaders;
+    try {
+      cliHeaders = await fs.readJSON(filePath);
+      // merge the headers defined with the CLI with those from the request
+      Object.assign(headers, cliHeaders);
+    } catch (error) {
+      log.error(`Failed to read or parse JSON file at ${filePath}: ${error.message}`);
```
Comment (Buuhuu) [inline src/server/HelixImportServer.js:77]: You probably don't need to catch the error here.

if reading the file fails, I as a developer would expect the cli to fail.

Comment (jckautzmann) [reply src/server/HelixImportServer.js]: good point, I removed the try-catch

File: src/import.js [hunk 2]
```diff
@@ -74,12 +74,18 @@ export default function up() {
           type: 'string',
           default: undefined,
         })
+        .option('headers-file', {
+          alias: 'headersFile',
+          describe: 'File location for your .json file for headers to be used in the requests',
```
Comment (kptdobe) [inline src/import.js:79]:
```suggestion
          describe: 'Location of a custom .json file containing headers to be used with all proxy requests',
```

**Review by kptdobe — CHANGES_REQUESTED** [review]

File: src/import.js (hunk 2 above)
Comment (jckautzmann) [reply src/import.js]: addressed in https://github.com/adobe/helix-cli/pull/2368/commits/0a07dc030893c6224008a1ec5007290065756363

File: src/import.cmd.js [hunk 3]
```diff
@@ -40,6 +40,11 @@ export default class ImportCommand extends AbstractServerCommand {
     return this;
   }
 
+  withHeadersFile(value) {
+    this._headersFile = value;
+    return this;
+  }
+
```
[the remark this answers is not in the document]
Comment (jckautzmann) [reply src/import.cmd.js:47]: `withHeadersFile` was already in `HelixImportProject` but I added a getter in `HelixImportProject`.
See https://github.com/adobe/helix-cli/pull/2368/commits/0a07dc030893c6224008a1ec5007290065756363

File: src/server/HelixImportServer.js [hunk 4]
```diff
@@ -63,6 +64,15 @@ export class HelixImportServer extends BaseServer {
     return url.toString();
   }
 
+  async _updateHeaders(headers) {
+    // eslint-disable-next-line no-underscore-dangle
+    const filePath = this._project._headersFile;
```
[the remark this answers is not in the document]
Comment (jckautzmann) [reply src/server/HelixImportServer.js]: addressed in https://github.com/adobe/helix-cli/pull/2368/commits/0a07dc030893c6224008a1ec5007290065756363

**Comment by jckautzmann** [comment]
@kptdobe I updated the PR based on your comment and added a few tests. Could you please review it and approve if everything is ok for you? Thx.

File: src/server/HelixImportServer.js [hunk 5]
```diff
@@ -63,6 +64,14 @@ export class HelixImportServer extends BaseServer {
     return url.toString();
   }
 
+  async _updateHeaders(headers) {
+    const filePath = this._project.headersFile;
+    if (!filePath) return;
```
Comment (tripodsan) [inline src/server/HelixImportServer.js:69]:
```suggestion
    if (!filePath) {
      return;
    }
```

File: src/server/HelixImportServer.js [hunk 6]
```diff
@@ -100,6 +109,7 @@ export class HelixImportServer extends BaseServer {
     delete headers.connection;
     delete headers.host;
     delete headers.referer;
+    await this._updateHeaders(headers);
```
Comment (tripodsan) [inline src/server/HelixImportServer.js:112]: you read the headers for each request? why don't you cache it on startup?

Comment (jckautzmann) [reply src/server/HelixImportServer.js:108]: Good point. I addressed it at https://github.com/adobe/helix-cli/pull/2368/commits/a302dc738a6a2a448b84521404f5bc32d2e32260

**Review by kptdobe — CHANGES_REQUESTED** [review]
Only small final nit: missing the test when the header file provided does not exist.

**Comment by jckautzmann** [comment]
I added the missing test at https://github.com/adobe/helix-cli/pull/2368/commits/87a6260a23c3d9bb1df2a0bdcea2e59e296513c8

**Review by kptdobe — APPROVED** [review]

File: src/server/HelixImportProject.js [hunk 7]
```diff
@@ -28,6 +29,20 @@ export class HelixImportProject extends BaseProject {
     return this._allowInsecure;
   }
 
+  withHeadersFile(value) {
+    this._headersFile = value;
+    this._cliHeaders = (value) ? fs.readJSON(value) : {};
```
Comment (tripodsan) [inline src/server/HelixImportProject.js:34]: that doesn't work. need too `await` the async call.

**Review by tripodsan — CHANGES_REQUESTED** [review]
I don't think this works without `awaiting` for the heads file to load

File: src/server/HelixImportProject.js (hunk 7 above)
Comment (kptdobe) [reply src/server/HelixImportProject.js:34]: Actually, that's a really good point.. why the tests are not capturing this ?

Comment (jckautzmann) [reply src/server/HelixImportProject.js]: With https://github.com/adobe/helix-cli/pull/2368/commits/00503e1e1d11d7fa0fe9e05b4b89c9e1e61fb7de the JSON file is read asynchronously

File: src/server/HelixImportProject.js [hunk 8]
```diff
@@ -28,6 +29,29 @@ export class HelixImportProject extends BaseProject {
...
+        this._cliHeaders = result;
+      }).catch((err) => {
+        this.log.error('error reading the headers file', err);
+        this._cliHeaders = {};
+      });
+    } else {
+      this._cliHeaders = {};
+    }
+    return this;
+  }
```
Comment (tripodsan) [inline src/server/HelixImportProject.js:45]: this won't work. you can't guarantee when the file has finished reading and the _cliHedaers is used before the callback is invoked.

you need to read the file during `init()`

**Review by tripodsan — CHANGES_REQUESTED** [review]
async problems

File: src/server/HelixImportProject.js (hunk 8 above)
Comment (jckautzmann) [reply src/server/HelixImportProject.js:45]: which `init()`? I don't see an `init()` in this file. Do you mean in the `async start()`?

Comment (tripodsan) [reply src/server/HelixImportProject.js:45]: yes. sorry, `start()`.

Comment (jckautzmann) [reply src/server/HelixImportProject.js]: I fixed the async issues when reading the headers file: https://github.com/adobe/helix-cli/pull/2368/commits/d1ebfc089d965608dae64b67728d5bbc924d5251
Also updated and added more tests.

**Review by tripodsan — APPROVED** [review]

**Comment by jckautzmann** [comment]
@kptdobe @tripodsan Now that the PR is approved, could you please merge it?

**Pull request merged**